cmd/cryptography/checksum.go

```diff
@@ -0,0 +1,246 @@
+package cryptography
+
+import (
+	"archive/zip"
+	"bufio"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	errs "github.com/open-cmsis-pack/cpackget/cmd/errors"
+	"github.com/open-cmsis-pack/cpackget/cmd/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+type checksum struct {
+	hashFunction, digest, filename string
+}
+
+func isValidHash(hashFunction string) bool {
+	valid := false
+	for _, h := range Hashes {
+		if h == hashFunction {
+			valid = true
+		}
+	}
+	return valid
+}
+
+// GetChecksumList receives a list of file paths and returns
+// their hashed content using either a specified function or
+// the default one
+func getChecksumList(baseDir string, filePathList []string, hashFunction string) ([]checksum, error) {
+	digests := make([]checksum, len(filePathList))
+	for i := 0; i < len(filePathList); i++ {
+		if utils.DirExists(filePathList[i]) {
+			continue
+		}
+		f, err := os.Open(filePathList[i])
+		if err != nil {
+			return []checksum{}, err
+		}
+		defer f.Close()
+
+		var h hash.Hash
+		switch hashFunction {
+		case "sha256":
+			h = sha256.New()
+			digests[i].hashFunction = "sha256"
+		default:
+			h = sha256.New()
+			digests[i].hashFunction = "sha256"
+		}
+
+		if _, err := io.Copy(h, f); err != nil {
+			return []checksum{}, err
+		}
+
+		digests[i].digest = fmt.Sprintf("%x", h.Sum(nil))
+		digests[i].filename = strings.ReplaceAll(filePathList[i], baseDir, "")
+	}
+	return digests, nil
+}
+
+func GenerateChecksum(sourcePack, destinationDir, hashFunction string) error {
+	// Use default Cryptographic Hash Function if none is provided
+	if hashFunction == "" {
+		hashFunction = Hashes[0]
+	}
+	if !isValidHash(hashFunction) {
+		return errs.ErrInvalidHashFunction
+	}
+
+	if !utils.FileExists(sourcePack) {
+		log.Errorf("\"%s\" is not a valid .pack", sourcePack)
```

Comment:
I think the message `not a valid .pack` is a bit confusing, the user might think the pack isn't valid, instead of not found.

Reply:
Done!